### PR TITLE
Houses and Nation Houses

### PR DESCRIPTION
**Pull request by GroobleDierne**
- `House` A collection of blocks forming a zone that can contain members & flags that apply to its members.
- `Nation House` A house that can be bought by a nation with a price associated and a free/occupied state.

- DivinePoints have also been renamed to DivineFavors in this PR.

### CODE REVIEW

File: src/main/java/net/laboulangerie/laboulangeriecore/houses/housewand/HouseWandCmd.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import net.kyori.adventure.text.Component;
+
+public class HouseWandCmd implements CommandExecutor {
+
+    private void giveWand(Player player) {
+        if (player == null) return;
+
+        ItemStack stack = new ItemStack(Material.IRON_AXE);
+        ItemMeta meta = stack.getItemMeta();
+        meta.setCustomModelData(69420);
```
Comment (PainOchoco) [inline src/main/java/net/laboulangerie/laboulangeriecore/houses/housewand/HouseWandCmd.java:21]: nice

Comment (GroobleDierne) [reply src/main/java/net/laboulangerie/laboulangeriecore/houses/housewand/HouseWandCmd.java:21]: ...

File: src/main/java/net/laboulangerie/laboulangeriecore/houses/housewand/HouseWandListener.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
+
+public class HouseWandListener implements Listener {
+
+    public static Location firstPos;
+    public static Location secondPos;
+
+    private int getBlocksCount() {
```
Comment (PainOchoco) [inline src/main/java/net/laboulangerie/laboulangeriecore/houses/housewand/HouseWandListener.java]: isn't it `abs(firstPos.x-secondPos.y)*abs(firstPos.y-secondPos.y)*abs(firstPos.z-secondPos.z)` ?

File: src/main/java/net/laboulangerie/laboulangeriecore/houses/nationhouse/HouseShop.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        if (event.getClick() != ClickType.LEFT || event.getCurrentItem() == null) return;
+
+        ItemStack item = event.getCurrentItem();
+        short page = Short.parseShort(title.split("page ")[1]);
+        Player player = (Player) event.getWhoClicked();
+
+        if (item.getType() == Material.ARROW) {
+            if (item.getItemMeta().getCustomModelData() == 2) { //Right arrow is clicked
+                displayShop(player, page); //No need for page++; because page's title is aldready increased by 1 to prevent displaying "page 0"
+            }else { //Left arrow is clicked (CustomModelData = 2)
```
Comment (PainOchoco) [inline src/main/java/net/laboulangerie/laboulangeriecore/houses/nationhouse/HouseShop.java]: CustomModelData != 2

**Review by PainOchoco — APPROVED** [review]
The count method can be simplified but except that it's pretty clean! good job

File: src/main/java/net/laboulangerie/laboulangeriecore/houses/nationhouse/HouseShop.java (hunk 3 above)
Comment (GroobleDierne) [reply src/main/java/net/laboulangerie/laboulangeriecore/houses/nationhouse/HouseShop.java]: The comment is wrong CustomModelData equals 1 for the left arrow

File: src/main/java/net/laboulangerie/laboulangeriecore/houses/HouseMembersCmd.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        return false;
+    }
+    @Override
+    public @Nullable List<String> onTabComplete(@NotNull CommandSender sender, @NotNull Command cmd, @NotNull String alias, @NotNull String[] args) {
+        List<String> suggestions = Arrays.asList("");
+
+        if (args.length == 1) {
+            suggestions = LaBoulangerieCore.housesManager.getHouses().values()
+                .stream().map(House::getName).collect(Collectors.toList());
+        }else if (args.length == 2) {
```
Comment (Blackoutburst) [inline src/main/java/net/laboulangerie/laboulangeriecore/houses/HouseMembersCmd.java]: Not really important but some spaces are missing on these }else

File: src/main/java/net/laboulangerie/laboulangeriecore/houses/housewand/HouseWandCmd.java (hunk 1 above)
Comment (Blackoutburst) [reply src/main/java/net/laboulangerie/laboulangeriecore/houses/housewand/HouseWandCmd.java:21]: Uh yes I can explain 😂

File: src/main/java/net/laboulangerie/laboulangeriecore/houses/nationhouse/NationHouseHolder.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    public void loadData() throws IOException {
+        if (!dataFile.exists()) {
+            dataFile.createNewFile();
+            return;
+        }
+
+        YamlConfiguration data = YamlConfiguration.loadConfiguration(dataFile);
+        if (data.getList("free-houses") != null) {
+            freeHouses = (List<UUID>) data.getStringList("free-houses").stream().map(UUID::fromString).collect(Collectors.toList());
+        }else data.createSection("free-houses");
```
Comment (Blackoutburst) [inline src/main/java/net/laboulangerie/laboulangeriecore/houses/nationhouse/NationHouseHolder.java]: Same for the `}else` maybe it's just how you code

File: src/main/java/net/laboulangerie/laboulangeriecore/houses/nationhouse/NationHousesCmd.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+                return true;
+            }
+            sender.sendMessage("§2-----§a[§epage " + (pageToDisplay) + "§a]§2-----");
+            if (args[1].equals("free")) {
+                for (int i = pageToDisplay*8; i < LaBoulangerieCore.nationHouseHolder.getFreeHouses().size() && i < (pageToDisplay+1)*8; i++) {
+                    UUID id = LaBoulangerieCore.nationHouseHolder.getFreeHouses().get(i);
+                    House house = LaBoulangerieCore.housesManager.getHouse(id);
+                    sender.sendMessage("§2" + house.getName() + " §r--- §3" + LaBoulangerieCore.nationHouseHolder.getHousePrice(id) + "$");
+                }
+            }else {
```
Comment (Blackoutburst) [inline src/main/java/net/laboulangerie/laboulangeriecore/houses/nationhouse/NationHousesCmd.java]: Same

File: src/main/java/net/laboulangerie/laboulangeriecore/misc/KingCondition.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import pl.betoncraft.betonquest.exceptions.QuestRuntimeException;
+
+public class KingCondition extends Condition {
+    @SuppressWarnings("deprecation")
+    public KingCondition(Instruction instruction) {
+        super(instruction);
+    }
+    @Override
+    protected Boolean execute(String playerID) throws QuestRuntimeException {
+        Resident resident = TownyUniverse.getInstance().getResident(UUID.fromString(playerID));
```
Comment (Blackoutburst) [inline src/main/java/net/laboulangerie/laboulangeriecore/misc/KingCondition.java:19]: Useless temp variable you could just `return TownyUniverse.getInstance().getResident(UUID.fromString(playerID)).isKing();`

**Review by Blackoutburst — APPROVED** [review]
Only a few really minor thing you don't even need to apply any changes good work as always GG

File: src/main/java/net/laboulangerie/laboulangeriecore/misc/KingCondition.java (hunk 7 above)
Comment (GroobleDierne) [reply src/main/java/net/laboulangerie/laboulangeriecore/misc/KingCondition.java:19]: In fact I should check if resident doesn't equal null

**Pull request merged**